### PR TITLE
Change XCom class methods to accept run_id argument

### PR DESCRIPTION
**Pull request by ashb**
Longer term (2.3 most likely) we will want to change the column on XCom
to have run_id instead of execution_date, but for now this changes the
interface to allow passing a run_id to XCom in place of an
execution_date.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/models/xcom.py [hunk 1]
```diff
@@ -55,6 +56,17 @@ class BaseXCom(Base, LoggingMixin):
     task_id = Column(String(ID_LEN, **COLLATION_ARGS), primary_key=True)
     dag_id = Column(String(ID_LEN, **COLLATION_ARGS), primary_key=True)
 
+    # For _now_, we link this via execution_date, in 2.3 we will migrate this table to use run_id too
+    dag_run = relationship(
+        "DagRun",
+        primaryjoin="""and_(
+            BaseXCom.dag_id == foreign(DagRun.dag_id),
+            BaseXCom.execution_date == foreign(DagRun.execution_date)
+        )""",
```
Comment (kaxil) [inline airflow/models/xcom.py:65]: Is this passed as string for late evaluation?

Comment (ashb) [reply airflow/models/xcom.py:65]: Yes, to avoid circular imports

File: airflow/models/xcom.py [hunk 2]
```diff
@@ -221,28 +267,51 @@ def delete(cls, xcoms, session=None):
...
+        :param run_id: Dag run id for the task
+        :type run_id: str or None
         :param session: database session
         :type session: sqlalchemy.orm.session.Session
         """
-        session.query(cls).filter(
+        # Given the historic order of this function (execution_date was first argument) to add a new optional
+        # param we need to add default values for everything :(
+        if not dag_id:
+            raise TypeError("clear() missing required argument: dag_id")
```
Comment (kaxil) [inline airflow/models/xcom.py:295]: Should this and L297 be `ValueError` too?

File: airflow/models/xcom.py [hunk 3]
```diff
@@ -68,12 +80,22 @@ def __repr__(self):
...
     @provide_session
-    def set(cls, key, value, execution_date, task_id, dag_id, session=None):
+    def set(cls, key, value, task_id, dag_id, execution_date=None, run_id=None, session=None):
         """
         Store an XCom value.
 
         :return: None
         """
+        if not (execution_date is None) ^ (run_id is None):
+            raise ValueError("Exactly one of execution_date or run_id must be passed")
```
Comment (uranusjr) [inline airflow/models/xcom.py:90]: I think argument count errors are usually TypeError?

File: airflow/models/xcom.py [hunk 4]
```diff
@@ -68,12 +80,22 @@ def __repr__(self):
...
+
+        if run_id:
+            from airflow.models.dagrun import DagRun
+
+            dag_run = session.query(DagRun).filter_by(dag_id=dag_id, run_id=run_id).one()
+
+            execution_date = dag_run.execution_date
+
         session.expunge_all()
```
Comment (uranusjr) [inline airflow/models/xcom.py:99]: 😞

File: airflow/models/xcom.py (hunk 2 above)
Comment (uranusjr) [reply airflow/models/xcom.py:295]: I think these should all be TypeError.

File: airflow/models/xcom.py (hunk 4 above)
Comment (ashb) [reply airflow/models/xcom.py:99]: Yuuuup. I wasn't touching it on purpose in this PR.

File: airflow/models/xcom.py (hunk 3 above)
Comment (ashb) [reply airflow/models/xcom.py:90]: There's no convention for this, but I went with ValueError here (but type error when missing) based on the description in the docs:

> Raised when an operation or function receives an argument that has the right type but an inappropriate value, and the situation is not described by a more precise exception such as IndexError.

And this from TypeError docs:

> Passing arguments of the wrong type (e.g. passing a list when an int is expected) should result in a TypeError, but passing arguments with the wrong value (e.g. a number outside expected boundaries) should result in a ValueError.

So given those two ValueError felt slightly more appropriate here. WDYT? I don't really mind either way

**Review by kaxil — APPROVED** [review]

**Pull request merged**